// @mui/material-next/node/Tabs/TabsList.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _styles = require("@mui/material/styles");
var _useTabsList = _interopRequireWildcard(require("@mui/base/useTabsList"));
var _TabsListContext = _interopRequireDefault(require("./TabsListContext"));
var _jsxRuntime = require("react/jsx-runtime");
const _excluded = ["variant", "indicator", "textColor", "children"];
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
const FlexContainer = (0, _styles.styled)('div', {
  name: 'MuiTabs',
  slot: 'FlexContainer',
  overridesResolver: (props, styles) => {
    const {
      ownerState
    } = props;
    return [styles.flexContainer, ownerState.vertical && styles.flexContainerVertical, ownerState.centered && styles.centered];
  }
})(({
  ownerState
}) => (0, _extends2.default)({
  display: 'flex'
}, ownerState.vertical && {
  flexDirection: 'column'
}, ownerState.centered && {
  justifyContent: 'center'
}));

/**
 * @ignore - internal component.
 */
const TabsList = /*#__PURE__*/React.forwardRef((props, ref) => {
  const {
      variant,
      indicator,
      textColor,
      children
    } = props,
    other = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);
  const {
    getRootProps,
    contextValue
  } = (0, _useTabsList.default)((0, _extends2.default)({}, props, {
    rootRef: ref
  }));
  const tabsListContextValue = React.useMemo(() => ({
    indicator,
    textColor,
    fullWidth: variant === 'fullWidth'
  }), [indicator, textColor, variant]);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_useTabsList.TabsListProvider, {
    value: contextValue,
    children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_TabsListContext.default.Provider, {
      value: tabsListContextValue,
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(FlexContainer, (0, _extends2.default)({}, other, getRootProps(), {
        children: children
      }))
    })
  });
});
process.env.NODE_ENV !== "production" ? TabsList.propTypes = {
  children: _propTypes.default.node,
  indicator: _propTypes.default.node,
  textColor: _propTypes.default.oneOf(['primary', 'secondary']),
  variant: _propTypes.default.oneOf(['fullWidth', 'scrollable', 'standard'])
} : void 0;
var _default = TabsList;
exports.default = _default;
